server/src/main/java/org/elasticsearch/index/mapper/ValueFetcher.java

```diff
@@ -50,4 +51,8 @@
      * Update the leaf reader used to fetch values.
      */
     default void setNextReader(LeafReaderContext context) {}
+
+    interface Context {
```

Comment:
I'm not sure that this is better than my last attempt. I think it's up to @jtibshirani to make the call on this one.

Reply:
I think that the difference is subtle, but I believe this should address the concerns around tracking usages. At the same time we are introducing a lightweight abstraction that clearly encapsulates what is needed to fetch values, which is much harder to read when functions are carried around. I would love for Julie to have a look and let us know what she thinks though.